learner/sampler/slurper.py

```diff
@@ -0,0 +1,252 @@
+        else:
+            self.unit = unit
+        
+
+    
+    def check_file_encodings(file: str = None, encoding: str = None, cycle_bit: int = 0) -> str:
+        """Ensure that the file can be opened by cycling through common file types."""        
+        if not file:
+            file = self.filename
+        if not encoding:
+            encoding = self.encoding            
+        
+        if cycle_bit > 0:
+            common_encodings = ["UTF-8","Latin-1", "UTF-16", "ascii", "cp037", "cp437", "UTF-32"]
+            return common_encodings[cycle_bit]
+            
+        else:
+            try:
+                with open(file, encoding = encoding) as f:
+                    f.seek(10000,0)
+                    f.readline()
+                    f.close()
+                return encoding
+            except:
+                common_encodings = ["Latin-1", "UTF-16", "ascii", "cp037", "cp437", "UTF-32"]
+
+                for codec in common_encodings:
+                    try:
+                        with open(file, encoding = codec) as f:
+                            f.readline()
+                            f.close()
+                        return codec
+                    except:
+                        continue
+                print("Your file is of unusual type - can you specify the encoding for us?")
+    
+    def estimate_line_size(file: str = None, encoding: str = None) -> int:
+        """Estimate the number of lines in the file- requires exactly 62 line reads."""
+        if not file:
+            file = self.filename
+        if not encoding:
+            encoding = self.encoding              
+        try:    
+            with open(file, encoding = encoding) as f:
+                #chop header
+                f.readline()
+                line = f.readline()
+                line_length = len(line)
+                f.seek(0,2)
+                eof = f.tell()
+                average_length = 0
+                for sample in range(2,32):
+                    f.seek(math.floor(nlines/sample),0)
+                    f.readline()
+                    line = f.readline()
+                    average_length += len(line)
+                f.close()
+        except:
+            for cycle in range(1,7):
+                new_encoding = self.check_file_encodings(file = file, encoding = None, cycle_bit = cycle)
+                try:
+                    with open(file, encoding = new_encoding) as f:
+                        #chop header
+                        f.readline()
+                        line = f.readline()
+                        line_length = len(line)
+                        f.seek(0,2)
+                        eof = f.tell()
+                        average_length = 0
+                        for sample in range(2,32):
+                            f.seek(math.floor(nlines/sample),0)
+                            f.readline()
+                            line = f.readline()
```

Comment:
I am 100% sure this could easily be broken down into multiple functions. Very hard to read and see what's going on because it's so long and convoluted. Also feel like there's unnecessary opening of files that could happen in a more systematic way.